src/coreclr/jit/utils.h

```diff
@@ -779,13 +780,128 @@ namespace CheckedOps
 const bool Unsigned = true;
 const bool Signed   = false;
 
-bool IntAddOverflows(int32_t firstAddend, int32_t secondAddend, bool unsignedAdd);
-bool IntSubOverflows(int32_t minuend, int32_t subtrahend, bool unsignedSub);
-bool IntMulOverflows(int32_t firstFactor, int32_t secondFactor, bool unsignedMul);
+// Important note: templated functions below must use dynamic "assert"s instead of "static_assert"s
+// because they can be instantiated on code paths that are not reachable at runtime, but visible
+// to the compiler. One example is VN's EvalOp<T> function, which can be instantiated with "size_t"
+// for some operators, and that's legal, but its callee EvalOpSpecialized<T> uses "assert(!AddOverflows(v1, v2))"
+// for VNF_ADD_OVF/UN, and would like to continue doing so without casts.
 
-bool LongAddOverflows(int64_t firstAddend, int64_t secondAddend, bool unsignedAddition);
-bool LongSubOverflows(int64_t minuend, int64_t subtrahend, bool unsignedSubtraction);
-bool LongMulOverflows(int64_t firstFactor, int64_t secondFactor, bool unsignedMultiplication);
+template <class T>
+bool AddOverflows(T firstAddend, T secondAddend, bool unsignedAdd)
+{
+    typedef typename std::make_unsigned<T>::type UT;
+    assert((std::is_same<T, int32_t>::value || std::is_same<T, int64_t>::value));
+
+    T result = firstAddend + secondAddend;
+
+    // For the SIGNED case - If there is one positive and one negative operand, there can be no overflow.
+    // If both are positive, the result has to be positive, and similary for negatives.
+    // For the UNSIGNED case - If an unsigned operand is bigger than the result then OVF.
+    if (unsignedAdd)
+    {
+        if ((static_cast<UT>(firstAddend) > static_cast<UT>(result)) ||
+            (static_cast<UT>(secondAddend) > static_cast<UT>(result)))
+        {
+            return true;
+        }
+    }
+    else if (((firstAddend < 0) == (secondAddend < 0)) && ((firstAddend < 0) != (result < 0)))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+template <class T>
+bool SubOverflows(T minuend, T subtrahend, bool unsignedSub)
```

Comment:
I think we could name parameters of these functions `left` and `right` (or `x` and `y`) and avoid rare-used words.

Reply:
I do not have a string opinion here. Renamed them to `x` and `y`.